File: Problems/Algorithms/802. Find Eventual Safe States/SafeStates.kt
class Solution {
    fun eventualSafeNodes(graph: Array<IntArray>): List<Int> {
        val n = graph.size
        val safe = BooleanArray(n) { false }
        
        val ngraph: MutableList<MutableList<Int>> = mutableListOf()
        val rgraph: MutableList<MutableList<Int>> = mutableListOf()
        for (i in 0..n-1) {
            ngraph.add(mutableListOf())
            rgraph.add(mutableListOf())
        }
        
        val queue = ArrayDeque<Int>()
        for (i in 0..n-1) {
            if (graph[i].size == 0) {
                queue.add(i)
            }
            
            for (j in graph[i]) {
                ngraph.get(i).add(j)
                rgraph.get(j).add(i)
            }
        }
        
        while (!queue.isEmpty()) {
            val i = queue.removeFirst()
            safe[i] = true
            
            for (j in rgraph.get(i)) {
                ngraph.get(j).remove(i)
                if (ngraph.get(j).isEmpty()) {
                    queue.add(j)
                }
            }
        }
        
        val results: MutableList<Int> = mutableListOf()
        for (i in 0..n-1) {
            if (safe[i]) results.add(i)
        }
        
        return results
    }
}

File: Problems/Algorithms/802. Find Eventual Safe States/safe-states.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        
        vector<int> indegree(n);
        vector<vector<int>> adj(n);
        for (int i = 0; i < n; i++) {
            for (auto node : graph[i]) {
                adj[node].push_back(i);
                indegree[i]++;
            }
        }

        queue<int> queue;
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                queue.push(i);
            }
        }

        vector<bool> safe(n);
        while (!queue.empty()) {
            int node = queue.front();
            queue.pop();
            safe[node] = true;

            for (auto& neighbor : adj[node]) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    queue.push(neighbor);
                }
            }
        }

        vector<int> ans;
        for (int i = 0; i < n; i++) {
            if (safe[i]) {
                ans.push_back(i);
            }
        }

        return ans;
    }
};
